File: packages/backend/tests/strategies.py
"""
Hypothesis strategies for generating test data for VyaparSaathi.

This module provides custom strategies for property-based testing using Hypothesis.
Strategies generate realistic sales data, festival calendars, and user inputs.
"""
from datetime import datetime, timedelta
from hypothesis import strategies as st
from typing import Any, Dict, List


# Business types
BUSINESS_TYPES = ["grocery", "apparel", "electronics", "general"]
STORE_SIZES = ["small", "medium", "large"]
REGIONS = ["north", "south", "east", "west", "central"]
RISK_TOLERANCES = ["conservative", "moderate", "aggressive"]
DATA_QUALITIES = ["poor", "fair", "good"]


@st.composite
def sales_record_strategy(draw) -> Dict[str, Any]:
    """
    Generate a valid sales record with required fields.
    
    Returns a dictionary with date, sku, and quantity fields.
    """
    # Generate date within last 2 years
    days_ago = draw(st.integers(min_value=0, max_value=730))
    date = datetime.now() - timedelta(days=days_ago)
    
    return {
        "date": date.strftime("%Y-%m-%d"),
        "sku": draw(st.text(alphabet=st.characters(whitelist_categories=("Lu", "Nd")), min_size=5, max_size=10)),
        "quantity": draw(st.integers(min_value=1, max_value=1000)),
        "category": draw(st.sampled_from(["grocery", "apparel", "electronics", "home", "beauty"])),
        "price": draw(st.floats(min_value=1.0, max_value=10000.0, allow_nan=False, allow_infinity=False)),
    }


@st.composite
def sales_data_list_strategy(draw, min_size: int = 1, max_size: int = 100) -> List[Dict[str, Any]]:
    """
    Generate a list of sales records.
    
    Args:
        min_size: Minimum number of records
        max_size: Maximum number of records
    """
    return draw(st.lists(sales_record_strategy(), min_size=min_size, max_size=max_size))


@st.composite
def invalid_sales_record_strategy(draw) -> Dict[str, Any]:
    """
    Generate an invalid sales record missing required fields or with invalid data.
    """
    record = {}
    
    # Randomly omit required fields
    if draw(st.booleans()):
        record["date"] = draw(st.one_of(
            st.just("invalid-date"),
            st.text(max_size=5),
            st.none(),
        ))
    
    if draw(st.booleans()):
        record["sku"] = draw(st.one_of(
            st.text(max_size=2),  # Too short
            st.none(),
        ))
    
    if draw(st.booleans()):
        record["quantity"] = draw(st.one_of(
            st.integers(max_value=0),  # Invalid quantity
            st.floats(allow_nan=True),
            st.none(),
        ))
    
    return record


@st.composite
def festival_event_strategy(draw) -> Dict[str, Any]:
    """
    Generate a festival event with all required fields.
    """
    festival_names = ["Diwali", "Eid", "Holi", "Christmas", "Pongal", "Onam", "Durga Puja"]
    
    # Generate date within next year
    days_ahead = draw(st.integers(min_value=1, max_value=365))
    date = datetime.now() + timedelta(days=days_ahead)
    
    return {
        "festivalId": f"{draw(st.sampled_from(festival_names)).lower()}-{date.year}",
        "name": draw(st.sampled_from(festival_names)),
        "date": date.strftime("%Y-%m-%d"),
        "region": draw(st.lists(st.sampled_from(REGIONS), min_size=1, max_size=3, unique=True)),
        "category": "festival",
        "demandMultipliers": {
            "grocery": draw(st.floats(min_value=1.0, max_value=5.0)),
            "apparel": draw(st.floats(min_value=1.0, max_value=5.0)),
            "electronics": draw(st.floats(min_value=1.0, max_value=5.0)),
        },
        "duration": draw(st.integers(min_value=1, max_value=10)),
        "preparationDays": draw(st.integers(min_value=7, max_value=30)),
    }


@st.composite
def user_profile_strategy(draw) -> Dict[str, Any]:
    """
    Generate a user profile with business information and preferences.
    """
    return {
        "userId": f"user-{draw(st.text(alphabet=st.characters(whitelist_categories=('Ll', 'Nd')), min_size=8, max_size=12))}",
        "businessInfo": {
            "name": draw(st.text(min_size=5, max_size=50)),
            "type": draw(st.sampled_from(BUSINESS_TYPES)),
            "location": {
                "city": draw(st.text(min_size=3, max_size=30)),
                "state": draw(st.text(min_size=3, max_size=30)),
                "region": draw(st.sampled_from(REGIONS)),
            },
            "size": draw(st.sampled_from(STORE_SIZES)),
        },
        "dataCapabilities": {
            "hasHistoricalData": draw(st.booleans()),
            "dataQuality": draw(st.sampled_from(DATA_QUALITIES)),
            "lastUpdated": datetime.now().isoformat(),
        },
        "preferences": {
            "forecastHorizon": draw(st.integers(min_value=7, max_value=14)),
            "riskTolerance": draw(st.sampled_from(RISK_TOLERANCES)),
        },
    }


@st.composite
def questionnaire_response_strategy(draw) -> Dict[str, Any]:
    """
    Generate a questionnaire response for low-data mode.
    """
    return {
        "userId": f"user-{draw(st.text(alphabet=st.characters(whitelist_categories=('Ll', 'Nd')), min_size=8, max_size=12))}",
        "businessType": draw(st.sampled_from(BUSINESS_TYPES)),
        "storeSize": draw(st.sampled_from(STORE_SIZES)),
        "lastFestivalPerformance": {
            "festival": draw(st.sampled_from(["Diwali", "Eid", "Holi", "Christmas"])),
            "salesIncrease": draw(st.floats(min_value=0.0, max_value=500.0)),
            "topCategories": draw(st.lists(st.sampled_from(["grocery", "apparel", "electronics"]), min_size=1, max_size=3, unique=True)),
            "stockoutItems": draw(st.lists(st.text(min_size=3, max_size=20), min_size=0, max_size=5)),
        },
        "currentInventory": draw(st.lists(
            st.fixed_dictionaries({
                "category": st.sampled_from(["grocery", "apparel", "electronics"]),
                "currentStock": st.integers(min_value=0, max_value=10000),
                "averageDailySales": st.integers(min_value=1, max_value=500),
                "confidence": st.sampled_from(["low", "medium", "high"]),
            }),
            min_size=1,
            max_size=5,
        )),
        "targetFestivals": draw(st.lists(st.sampled_from(["Diwali", "Eid", "Holi", "Christmas"]), min_size=1, max_size=3, unique=True)),
    }


@st.composite
def forecast_request_strategy(draw) -> Dict[str, Any]:
    """
    Generate a forecast request.
    """
    return {
        "userId": f"user-{draw(st.text(alphabet=st.characters(whitelist_categories=('Ll', 'Nd')), min_size=8, max_size=12))}",
        "forecastHorizon": draw(st.integers(min_value=7, max_value=14)),
        "targetFestivals": draw(st.lists(st.text(min_size=3, max_size=20), min_size=0, max_size=3)),
        "dataMode": draw(st.sampled_from(["structured", "low-data"])),
        "confidence": draw(st.floats(min_value=0.0, max_value=1.0)),
    }


@st.composite
def inventory_data_strategy(draw) -> Dict[str, Any]:
    """
    Generate inventory data for risk assessment.
    """
    return {
        "sku": draw(st.text(alphabet=st.characters(whitelist_categories=("Lu", "Nd")), min_size=5, max_size=10)),
        "category": draw(st.sampled_from(["grocery", "apparel", "electronics"])),
        "currentStock": draw(st.integers(min_value=0, max_value=10000)),
        "reorderPoint": draw(st.integers(min_value=10, max_value=500)),
        "leadTimeDays": draw(st.integers(min_value=1, max_value=30)),
    }


@st.composite
def daily_prediction_strategy(draw) -> Dict[str, Any]:
    """
    Generate a daily prediction for forecast results.
    """
    days_ahead = draw(st.integers(min_value=0, max_value=14))
    date = datetime.now() + timedelta(days=days_ahead)
    
    demand_forecast = draw(st.floats(min_value=1.0, max_value=1000.0, allow_nan=False, allow_infinity=False))
    festival_multiplier = draw(st.floats(min_value=1.0, max_value=5.0, allow_nan=False, allow_infinity=False))
    
    # Calculate bounds around forecast
    variance = draw(st.floats(min_value=0.1, max_value=0.5))
    lower_bound = max(0, demand_forecast * (1 - variance))
    upper_bound = demand_forecast * (1 + variance)
    
    return {
        "date": date.strftime("%Y-%m-%d"),
        "demandForecast": demand_forecast,
        "lowerBound": lower_bound,
        "upperBound": upper_bound,
        "festivalMultiplier": festival_multiplier,
        "confidence": draw(st.floats(min_value=0.5, max_value=0.95)),
        "festivals": draw(st.lists(st.sampled_from(["Diwali", "Eid", "Holi", "Christmas"]), max_size=2))
    }


@st.composite
def forecast_result_strategy(draw) -> Dict[str, Any]:
    """
    Generate a complete forecast result for testing explanations.
    """
    predictions = draw(st.lists(daily_prediction_strategy(), min_size=7, max_size=14))
    
    return {
        "sku": draw(st.text(alphabet=st.characters(whitelist_categories=("Lu", "Nd")), min_size=5, max_size=10)),
        "category": draw(st.sampled_from(["grocery", "apparel", "electronics", "home", "beauty"])),
        "predictions": predictions,
        "confidence": draw(st.floats(min_value=0.5, max_value=0.95)),
        "methodology": draw(st.sampled_from(["ml", "pattern", "hybrid"])),
        "assumptions": draw(st.lists(st.text(min_size=10, max_size=100), min_size=1, max_size=5))
    }


@st.composite
def risk_assessment_strategy(draw) -> Dict[str, Any]:
    """
    Generate a risk assessment for testing explanations.
    """
    current_stock = draw(st.integers(min_value=0, max_value=10000))
    
    stockout_prob = draw(st.floats(min_value=0.0, max_value=1.0))
    overstock_prob = draw(st.floats(min_value=0.0, max_value=1.0))
    
    return {
        "sku": draw(st.text(alphabet=st.characters(whitelist_categories=("Lu", "Nd")), min_size=5, max_size=10)),
        "category": draw(st.sampled_from(["grocery", "apparel", "electronics", "home", "beauty"])),
        "currentStock": current_stock,
        "stockoutRisk": {
            "probability": stockout_prob,
            "daysUntilStockout": draw(st.integers(min_value=0, max_value=30)),
            "potentialLostSales": draw(st.floats(min_value=0.0, max_value=5000.0))
        },
        "overstockRisk": {
            "probability": overstock_prob,
            "excessUnits": draw(st.integers(min_value=0, max_value=5000)),
            "carryingCost": draw(st.floats(min_value=0.0, max_value=10000.0))
        },
        "recommendation": {
            "action": draw(st.sampled_from(["reorder", "reduce", "maintain"])),
            "suggestedQuantity": draw(st.integers(min_value=0, max_value=5000)),
            "urgency": draw(st.sampled_from(["low", "medium", "high"])),
            "reasoning": draw(st.lists(st.text(min_size=10, max_size=100), min_size=1, max_size=5)),
            "confidence": draw(st.floats(min_value=0.5, max_value=0.95))
        }
    }

File: packages/frontend/src/test/strategies.ts
/**
 * Fast-check strategies for generating test data for VyaparSaathi frontend.
 * 
 * This module provides custom arbitraries for property-based testing using fast-check.
 * Strategies generate realistic sales data, festival calendars, and user inputs.
 */

import * as fc from 'fast-check';

// Type definitions
export interface SalesRecord {
  date: string;
  sku: string;
  quantity: number;
  category?: string;
  price?: number;
}

export interface FestivalEvent {
  festivalId: string;
  name: string;
  date: string;
  region: string[];
  category: string;
  demandMultipliers: Record<string, number>;
  duration: number;
  preparationDays: number;
}

export interface UserProfile {
  userId: string;
  businessInfo: {
    name: string;
    type: string;
    location: {
      city: string;
      state: string;
      region: string;
    };
    size: string;
  };
  dataCapabilities: {
    hasHistoricalData: boolean;
    dataQuality: string;
    lastUpdated: string;
  };
  preferences: {
    forecastHorizon: number;
    riskTolerance: string;
  };
}

export interface ForecastRequest {
  userId: string;
  forecastHorizon: number;
  targetFestivals: string[];
  dataMode: 'structured' | 'low-data';
  confidence: number;
}

export interface DailyPrediction {
  date: string;
  demandForecast: number;
  lowerBound: number;
  upperBound: number;
  festivalMultiplier: number;
}

// Constants
const BUSINESS_TYPES = ['grocery', 'apparel', 'electronics', 'general'] as const;
const STORE_SIZES = ['small', 'medium', 'large'] as const;
const REGIONS = ['north', 'south', 'east', 'west', 'central'] as const;
const RISK_TOLERANCES = ['conservative', 'moderate', 'aggressive'] as const;
const DATA_QUALITIES = ['poor', 'fair', 'good'] as const;
const CATEGORIES = ['grocery', 'apparel', 'electronics', 'home', 'beauty'] as const;
const FESTIVAL_NAMES = ['Diwali', 'Eid', 'Holi', 'Christmas', 'Pongal', 'Onam', 'Durga Puja'] as const;

/**
 * Generate a valid sales record with required fields.
 */
export const salesRecordArbitrary = (): fc.Arbitrary<SalesRecord> => {
  return fc.record({
    date: fc.date({ min: new Date('2021-01-01'), max: new Date() }).map(d => d.toISOString().split('T')[0]),
    sku: fc.stringMatching(/^SKU[A-Z0-9]{5,10}$/),
    quantity: fc.integer({ min: 1, max: 1000 }),
    category: fc.constantFrom(...CATEGORIES),
    price: fc.float({ min: 1.0, max: 10000.0, noNaN: true }),
  });
};

/**
 * Generate a list of sales records.
 */
export const salesDataListArbitrary = (minSize = 1, maxSize = 100): fc.Arbitrary<SalesRecord[]> => {
  return fc.array(salesRecordArbitrary(), { minLength: minSize, maxLength: maxSize });
};

/**
 * Generate an invalid sales record missing required fields or with invalid data.
 */
export const invalidSalesRecordArbitrary = (): fc.Arbitrary<Partial<SalesRecord>> => {
  return fc.oneof(
    // Missing date
    fc.record({
      sku: fc.string({ minLength: 5, maxLength: 10 }),
      quantity: fc.integer({ min: 1, max: 1000 }),
    }),
    // Missing sku
    fc.record({
      date: fc.date().map(d => d.toISOString().split('T')[0]),
      quantity: fc.integer({ min: 1, max: 1000 }),
    }),
    // Missing quantity
    fc.record({
      date: fc.date().map(d => d.toISOString().split('T')[0]),
      sku: fc.string({ minLength: 5, maxLength: 10 }),
    }),
    // Invalid quantity (negative or zero)
    fc.record({
      date: fc.date().map(d => d.toISOString().split('T')[0]),
      sku: fc.string({ minLength: 5, maxLength: 10 }),
      quantity: fc.integer({ max: 0 }),
    }),
    // Invalid date format
    fc.record({
      date: fc.string({ maxLength: 5 }),
      sku: fc.string({ minLength: 5, maxLength: 10 }),
      quantity: fc.integer({ min: 1, max: 1000 }),
    }),
  );
};

/**
 * Generate a festival event with all required fields.
 */
export const festivalEventArbitrary = (): fc.Arbitrary<FestivalEvent> => {
  return fc.record({
    festivalId: fc.constantFrom(...FESTIVAL_NAMES).chain(name =>
      fc.integer({ min: 2023, max: 2025 }).map(year => `${name.toLowerCase()}-${year}`)
    ),
    name: fc.constantFrom(...FESTIVAL_NAMES),
    date: fc.date({ min: new Date(), max: new Date(Date.now() + 365 * 24 * 60 * 60 * 1000) })
      .map(d => d.toISOString().split('T')[0]),
    region: fc.array(fc.constantFrom(...REGIONS), { minLength: 1, maxLength: 3 }).map(arr => [...new Set(arr)]),
    category: fc.constant('festival'),
    demandMultipliers: fc.record({
      grocery: fc.float({ min: 1.0, max: 5.0, noNaN: true }),
      apparel: fc.float({ min: 1.0, max: 5.0, noNaN: true }),
      electronics: fc.float({ min: 1.0, max: 5.0, noNaN: true }),
    }),
    duration: fc.integer({ min: 1, max: 10 }),
    preparationDays: fc.integer({ min: 7, max: 30 }),
  });
};

/**
 * Generate a user profile with business information and preferences.
 */
export const userProfileArbitrary = (): fc.Arbitrary<UserProfile> => {
  return fc.record({
    userId: fc.hexaString({ minLength: 8, maxLength: 12 }).map(s => `user-${s}`),
    businessInfo: fc.record({
      name: fc.string({ minLength: 5, maxLength: 50 }),
      type: fc.constantFrom(...BUSINESS_TYPES),
      location: fc.record({
        city: fc.string({ minLength: 3, maxLength: 30 }),
        state: fc.string({ minLength: 3, maxLength: 30 }),
        region: fc.constantFrom(...REGIONS),
      }),
      size: fc.constantFrom(...STORE_SIZES),
    }),
    dataCapabilities: fc.record({
      hasHistoricalData: fc.boolean(),
      dataQuality: fc.constantFrom(...DATA_QUALITIES),
      lastUpdated: fc.date().map(d => d.toISOString()),
    }),
    preferences: fc.record({
      forecastHorizon: fc.integer({ min: 7, max: 14 }),
      riskTolerance: fc.constantFrom(...RISK_TOLERANCES),
    }),
  });
};

/**
 * Generate a forecast request.
 */
export const forecastRequestArbitrary = (): fc.Arbitrary<ForecastRequest> => {
  return fc.record({
    userId: fc.hexaString({ minLength: 8, maxLength: 12 }).map(s => `user-${s}`),
    forecastHorizon: fc.integer({ min: 7, max: 14 }),
    targetFestivals: fc.array(fc.string({ minLength: 3, maxLength: 20 }), { maxLength: 3 }),
    dataMode: fc.constantFrom('structured', 'low-data'),
    confidence: fc.float({ min: 0.0, max: 1.0, noNaN: true }),
  });
};

/**
 * Generate a daily prediction.
 */
export const dailyPredictionArbitrary = (): fc.Arbitrary<DailyPrediction> => {
  return fc.record({
    date: fc.date({ min: new Date(), max: new Date(Date.now() + 14 * 24 * 60 * 60 * 1000) })
      .map(d => d.toISOString().split('T')[0]),
    demandForecast: fc.float({ min: 0, max: 1000, noNaN: true }),
    lowerBound: fc.float({ min: 0, max: 500, noNaN: true }),
    upperBound: fc.float({ min: 500, max: 1500, noNaN: true }),
    festivalMultiplier: fc.float({ min: 1.0, max: 5.0, noNaN: true }),
  });
};

/**
 * Generate inventory data for risk assessment.
 */
export const inventoryDataArbitrary = () => {
  return fc.record({
    sku: fc.stringMatching(/^SKU[A-Z0-9]{5,10}$/),
    category: fc.constantFrom(...CATEGORIES),
    currentStock: fc.integer({ min: 0, max: 10000 }),
    reorderPoint: fc.integer({ min: 10, max: 500 }),
    leadTimeDays: fc.integer({ min: 1, max: 30 }),
  });
};
